indigo/analysis/refs/base.py

```diff
@@ -123,3 +123,121 @@ class RefsFinderENG(BaseRefsFinder):
             )
         ''', re.X)
     candidate_xpath = ".//text()[contains(., 'Act') and not(ancestor::a:ref)]"
+
+
+class BaseSectionRefsFinder(LocaleBasedMatcher):
+    """ Finds internal references to sections in documents.
+
+    """
+
+    section_re = None
+    """ This must be defined by a subclass. It should be a compiled regular
+    expression, with named captures for `ref` and `num`.
+    """
+    # TODO: add subsection reference
+    candidate_xpath = None  # this must be defined by a subclass
+
+    # the ancestor elements that can contain references
+    ancestors = ['body', 'mainBody', 'conclusions']
+    # TODO: check on ancestor elements
+
+    def find_references_in_document(self, document):
+        """ Find references in +document+, which is an Indigo Document object.
+        """
+        # we need to use etree, not objectify, so we can't use document.doc.root,
+        # we have to re-parse it
+        root = etree.fromstring(document.content)
+        self.setup(root)
+        self.find_references(root)
+        document.content = etree.tostring(root, encoding='utf-8').decode('utf-8')
+
+    def setup(self, root):
+        self.ns = root.nsmap[None]
+        self.nsmap = {'a': self.ns}
+        self.ref_tag = "{%s}ref" % self.ns
+        self.ancestor_xpath = etree.XPath('|'.join(f'.//a:{a}' for a in self.ancestors), namespaces=self.nsmap)
+        self.candidate_xpath = etree.XPath(self.candidate_xpath, namespaces=self.nsmap)
+
+    def make_href(self, node, match):
+        """ Turn this match into an href
+        """
+        num = match.group("num")
```

Comment:
Technically you should escape any possible `'` chars in `num`, otherwise they will break the xpath. But the chances of the num having an apostrophe are pretty slim.

Reply:
thanks, that's good to bear in mind! 
the regex shouldn't allow for accidentally including apostrophes though (especially not inside `num`), so I'll just file this away for future reference if that's good with you?

Reply:
sure. it'll break obviously if it breaks, which is better than quietly. :)